src/nxt_conf_validation.c

```diff
@@ -1551,7 +1551,7 @@ static nxt_int_t
 nxt_conf_vldt_var(nxt_conf_validation_t *vldt, const nxt_str_t *name,
     nxt_str_t *value)
 {
-    u_char  error[NXT_MAX_ERROR_STR];
+    u_char  error[NXT_MAX_ERROR_STR] = {0};
 
     if (nxt_tstr_test(vldt->tstr_state, value, error) != NXT_OK) {
         return nxt_conf_vldt_error(vldt, "%s in the \"%V\" value.",
```

Comment:
Unless you have any more details, this looks like a false positive.

```c
    if (nxt_tstr_test(vldt->tstr_state, value, error) != NXT_OK) {              
        return nxt_conf_vldt_error(vldt, "%s in the \"%V\" value.",             
                                   error, name);                                
    }
```

When we return from `nxt_tstr_test()` there are two possibilities

1) The return value is NXT_OK & error hasn't been set.
2) The return value is NXT_ERROR & error **has** been set.

In the case of (2) we then format the error message.

Reply:
I know some simple things: static analyzers have false positives, dynamic analyzers don't. This is why I don't use static analyzers except for Cyclomatic Complexity.

You are right, the provided patch  https://github.com/nginx/unit/pull/1327/commits/576a18b70b43deb95a38d7bfd1ecb29132a60d5a is wrong.

I did some debugging,
When `nxt_tstr_test` is called, `nxt_tstr_test` calls `nxt_sprintf`.
```c
nxt_int_t
nxt_tstr_test(nxt_tstr_state_t *state, nxt_str_t *str, u_char *error)
{
    //...
#else
        nxt_sprintf(error, error + NXT_MAX_ERROR_STR,
                    "Unit is built without support of njs: "
                    "\"--njs\" ./configure option is missing.");
        return NXT_ERROR;
#endif

  //...

}
```

Which ends up calling `nxt_vsprintf`, This is the only code that is executed.
```c
u_char *
nxt_vsprintf(u_char *buf, u_char *end, const char *fmt, va_list args)
{
    // ...

    while (*fmt != '\0' && buf < end) {

        /*
         * "buf < end" means that we could copy at least one character:
         * a plain character, "%%", "%c", or a minus without test.
         */

        if (*fmt != '%') {
            *buf++ = *fmt++;
            continue;
        }


    // ...

    }

    return buf;
}
```

The `while` loop copies everything except `'\0'`.

Reply:
### Patch to reproduce in controlled environment.

Well, I first thought I could be wrong, so I did some tests.

```patch
diff --git a/fuzzing/build-fuzz.sh b/fuzzing/build-fuzz.sh
index 04f080d9..2a11870e 100644
--- a/fuzzing/build-fuzz.sh
+++ b/fuzzing/build-fuzz.sh
@@ -2,8 +2,8 @@
 
 export CC=clang
 export CXX=clang++
-export CFLAGS="-g -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address,undefined -fsanitize=fuzzer-no-link"
-export CXXFLAGS="-g -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address,undefined -fsanitize=fuzzer-no-link"
+export CFLAGS="-O0 -fno-omit-frame-pointer -gline-tables-only -Wno-error=enum-constexpr-conversion -Wno-error=incompatible-function-pointer-types -Wno-error=int-conversion -Wno-error=deprecated-declarations -Wno-error=implicit-function-declaration -Wno-error=implicit-int -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=memory -fsanitize-memory-track-origins -fsanitize=fuzzer-no-link"
+export CXXFLAGS="-O0 -fno-omit-frame-pointer -gline-tables-only -Wno-error=enum-constexpr-conversion -Wno-error=incompatible-function-pointer-types -Wno-error=int-conversion -Wno-error=deprecated-declarations -Wno-error=implicit-function-declaration -Wno-error=implicit-int -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=memory -fsanitize-memory-track-origins -fsanitize=fuzzer-no-link -stdlib=libc++"
 export LIB_FUZZING_ENGINE="-fsanitize=fuzzer"
 
 ./configure --no-regex --no-pcre2 --fuzz=$LIB_FUZZING_ENGINE
diff --git a/src/nxt_sprintf.c b/src/nxt_sprintf.c
index 875f43a5..a606301f 100644
--- a/src/nxt_sprintf.c
+++ b/src/nxt_sprintf.c
@@ -160,6 +160,18 @@ nxt_vsprintf(u_char *buf, u_char *end, const char *fmt, va_list args)
                 continue;
             }
 
+            /*
+             * Just lazy work.
+             */
+            printf("\nPrint p at 75 as hax : %02x", (unsigned char)p[75]);
+            printf("\nPrint p at 75 as char: %c",   (unsigned char)p[75]);
+            printf("\nPrint value of error in nxt_vsprintf as p: %s \n", p);
+
+            /*
+             * There is no byte, this should crash here.
+             */
+            printf("\nPrint p at 76 as hax : %02x", (unsigned char)p[76]);
+
             while (*p != '\0' && buf < end) {
                 *buf++ = *p++;
             }
```

[clusterfuzz-testcase-minimized-fuzz_json-5545917827055616.txt](https://github.com/user-attachments/files/15935479/clusterfuzz-testcase-minimized-fuzz_json-5545917827055616.txt)

Reply:
Just for fun, you can try adding this hacky line in `nxt_conf_validation.c` to see if everything works as expected afterward.

```patch
--- a/src/nxt_conf_validation.c
+++ b/src/nxt_conf_validation.c
@@ -1554,6 +1554,7 @@ nxt_conf_vldt_var(nxt_conf_validation_t *vldt, const nxt_str_t *name,
     u_char  error[NXT_MAX_ERROR_STR] = {0};
 
     if (nxt_tstr_test(vldt->tstr_state, value, error) != NXT_OK) {
+        error[76] = '\0';
         return nxt_conf_vldt_error(vldt, "%s in the \"%V\" value.",
                                    error, name);
     }
```

Reply:
I think it would be better if I remove https://github.com/nginx/unit/pull/1327/commits/576a18b70b43deb95a38d7bfd1ecb29132a60d5a from this PR. You can fix it in a separate PR with `Issue: 69744`.

Reply:
Hi @pkillarjun,
Good catch. It's an issue that missed `\0` at the end of `nxt_sprintf()` in `nxt_tstr_test()`.
Your way to fix it is correct, but I'd suggest fixing it like this:
```
diff --git a/src/nxt_tstr.c b/src/nxt_tstr.c
index fde4822d..6f39cff2 100644
--- a/src/nxt_tstr.c
+++ b/src/nxt_tstr.c
@@ -159,7 +159,7 @@ nxt_tstr_test(nxt_tstr_state_t *state, nxt_str_t *str, u_char *error)
 #else
         nxt_sprintf(error, error + NXT_MAX_ERROR_STR,
                     "Unit is built without support of njs: "
-                    "\"--njs\" ./configure option is missing.");
+                    "\"--njs\" ./configure option is missing.%Z");
         return NXT_ERROR;
 #endif
```
There are more than one places to use `NXT_MAX_ERROR_STR`, and it's better to keep it just one way.
We usually add `%Z` on demand. In this case, it's an issue since it's used in the format `"%s in this \"%V\" value."`.
Thanks.

Reply:
`u_char  error[NXT_MAX_ERROR_STR] = {0};` This only fixes the symptoms, not the actual problem.

Reply:
Hmm, I mean we don't need to set it zero but append `%z` instead.

Reply:
Oh, sorry, missed it. I thought `%z` was for extra formatting.
I thought it was a note to yourself to add extra features instead of fixing this bug.

Reply:
Correct.